Post processor

Proca-Consul opened the pull request:
The first block (i.e. the aggregator) has been implemented.
I wrote two tests, both passing. Should you have any idea for some additional one, just fire away.

#domenicoMortacciTua

giuliojiang commented on backend/postprocessor/src/test/java/postprocessor/AggregatorTest.java:
```diff
@@ -0,0 +1,46 @@
+    public void producesCorrectlyArrangedJSONFileFromAggregationOfTwoOutputFiles() {
+
+        List<String> paths = new ArrayList<>();
+        paths.add("src/test/testFiles/msOutputFile1.json");
```
sure assumptions about the starting working directory don't break these relative paths?

Proca-Consul replied:
What do you mean exactly? :)

helicopter88 replied:
The test files may not be found if the working directory changes

Proca-Consul replied:
Right ok I see the problem. How can I go about this? :)

AbraaoMota replied:
These work with make - maven takes care of relative dirs

AbraaoMota commented on backend/postprocessor/src/main/java/postprocessor/PostProcessor.java:
```diff
@@ -0,0 +1,31 @@
+        Aggregator aggregator = new Aggregator(inputJsonPaths);
+        String jsonAggregatedErrors = aggregator.aggregate();
+
+
```
whitespace

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,98 @@
+     * @param map
+     * @param annotations
+     */
+    private void addAnnotationToMapByErrorType(Map<ErrorType, List<JSONObject>> map, JSONArray annotations) {
```
Name seems a bit too long

Proca-Consul replied:
It's self explanatory and does not harm (in fact it does the opposite of harm.)

helicopter88 replied:
Just noticed, JavaDoc is ill-formed.

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/ErrorType.java:
```diff
@@ -0,0 +1,22 @@
+package postprocessor;
+
+public enum ErrorType {
+    SYNTAX("syntax", 10.0),
```
I think this is wrong, if you get a syntax error, you should cap the mark at E, as that's the DoC policy

Proca-Consul replied:
It just a dummy value for the moment, given we did not reach any official agreement on this. What you say is completely reasonable anyways

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/ErrorType.java:
```diff
@@ -0,0 +1,22 @@
+package postprocessor;
```
Do we really need to recalculate the score given that all the microservices are giving you a score back?

AbraaoMota replied:
for now I think we can just produce a microservice average but if we want to do ML over the microservice importance to final grades we probably want to implement a score recalculator in the post-processor

AbraaoMota commented on backend/postprocessor/src/test/testFiles/aggregationOf1and2.json:
```diff
@@ -0,0 +1,43 @@
+    {
+      "charno": 22,
+      "errortype": "semantic",
+      "text": "The eyes are the nipples of the face.",
```
Maybe not have this ...

Proca-Consul replied:
This is the output we expect. We compare the generated output against this.

AbraaoMota replied:
I mean change the actual text..

Proca-Consul replied:
There can be ANYTHING.

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessor.java:
```diff
@@ -0,0 +1,31 @@
+    public static void main(String[] args) {
+
+        if (args.length <= 1) {
+            throw new IllegalArgumentException(
```
Too long, maybe split to new lines

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessor.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static void main(String[] args) {
+
+        if (args.length <= 1) {
```
Shouldn't this be <= 3?

Proca-Consul replied:
Correct. @AbraaoMota

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessor.java:
```diff
@@ -0,0 +1,31 @@
+
+        Aggregator aggregator = new Aggregator(inputJsonPaths);
+        String jsonAggregatedErrors = aggregator.aggregate();
+
```
Useless empty line

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessorUtils.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            Object obj = parser.parse(new FileReader(pathToFile));
+            output = (JSONObject) obj;
+        } catch (IOException e) {
```
Multicatch using |

Proca-Consul replied:
It's java 7 <3

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessorUtils.java:
```diff
@@ -0,0 +1,55 @@
+
+    /**
+     * Get JSON file from path
+     * @param pathToFile
```
Return what?

helicopter88 replied:
Well the documentation is ill formed if you don't specify that, and mvn site may not compile

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/PostProcessorUtils.java:
```diff
@@ -0,0 +1,55 @@
+            case "style":
+                errorType = ErrorType.STYLE;
+                break;
+        }
```
No default case?

Proca-Consul replied:
Good one. @AbraaoMota can you add it?

helicopter88 commented on backend/postprocessor/src/test/testFiles/aggregationOf1and2.json:
```diff
@@ -0,0 +1,43 @@
+    {
+      "charno": 61,
+      "errortype": "syntax",
+      "text": "READ THE EMAIL PLEASE. --Bryan Mitchell",
```
This is an assessed exercise, I would rather not add jokes in it

AbraaoMota replied:
+1, this should go for the other mention in the other document in documentation

Proca-Consul replied:
It just copied from Giulio's document. Also, they are test files. It's ok :)

mihaicar replied:
even his name is wrong so... i agree with helicopter

helicopter88 commented on backend/postprocessor/src/test/testFiles/aggregationOf1and2.json:
```diff
@@ -0,0 +1,43 @@
+      "charno": 22,
+      "errortype": "semantic",
+      "text": "The eyes are the nipples of the face.",
+      "lineno": 53,
```
????

Proca-Consul replied:
Same

helicopter88 commented on backend/postprocessor/src/test/testFiles/msOutputFile1.json:
```diff
@@ -0,0 +1,20 @@
+      "filename": "quadratic.hs",
+      "lineno": 19,
+      "charno": 61,
+      "text": "READ THE EMAIL PLEASE. --Bryan Mitchell"
```
Again

Proca-Consul replied:
Same

helicopter88 commented on backend/postprocessor/src/test/testFiles/msOutputFile2.json:
```diff
@@ -0,0 +1,27 @@
+      "filename": "quadratic.hs",
+      "lineno": 53,
+      "charno": 22,
+      "text": "The eyes are the nipples of the face."
```
??????

Proca-Consul replied:
Same

helicopter88 commented on backend/postprocessor/src/test/java/postprocessor/AggregatorTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Test
+    public void skipsMicroServiceOutputIfAnyErrorIsOccurredInTheMicroservices() {
```
This name seems a bit too long, have you thought about using specs so that it looks much nicer?

Proca-Consul replied:
It's the practice we learned from SED :)

helicopter88 replied:
I think for long names, the frameworks that use specs look better (check the scala tests that I have) and are also more readable, after all this is more BDD than TDD, so you want to specify the behaviour

mihaicar replied:
also, besides the long name, consistency - MicroService or Microservices. i know it's a petty thing, but...

Proca-Consul commented on backend/postprocessor/src/test/testFiles/msOutputFile1.json:
```diff
@@ -0,0 +1,20 @@
+    }
+  ],
+  "errors": []
+}
```
They are not, they are the crafted "arguments" for the aggregator.

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/ErrorType.java:
```diff
@@ -0,0 +1,22 @@
+    SYNTAX("syntax", 10.0),
+    SEMANTIC("semantic", 1.0),
+    STYLE("style", 0.1);
+
```
Also, what about having a matcher method here instead of the case switch?

helicopter88 replied:
the case switch is further down in this CR, you can just specify a string version of each enum value and have a public ErrorType(String string) that returns the correct one

helicopter88 commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+
```
White space

Proca-Consul commented:
@giuliojiang As for the```make test``` I am not sure, I did not take care of that explicitly. I guess it works by modifying the Makefile in order to execute these tests too right?

helicopter88 commented:
If your tests work fine when invoking mvn test, they will work fine when invoking make test

helicopter88 commented:
Commit names shouldn't refer to stuff you haven't touched such as the .class field..

helicopter88 commented:
And also, get rid of the Merge commit.

Proca-Consul commented:
@helicopter88  I don't get the commit name thing at all. Commit names are summary of what you have done, and I removed the .class files. Yesterday I made a bit of a mess committing like crazy from the GUI (reason why we have 2374582905 commits.) I squashed them but don't know how to remove the single commits from here. 

As for the author in the commit name, we should have discussed this beforehand. It's a good practice I took last here that was much appreciated by Mark :) But I can just not do it for this project

Proca-Consul commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,98 @@
+
+        Map<ErrorType, List<JSONObject>> map = new HashMap<>();
+
+        // Create massive list containing all errors
```
It's not labelled, is a comment :)

helicopter88 replied:
Still, you're implying that the list is always massive(and in that case you may want to find the most optimal structure for them (e.g. a tree based structure if they have to be ordered)), since we do not care about that, just do not give it any unneeded attribute.

Proca-Consul commented on backend/pom.xml:
```diff
@@ -1,5 +1,5 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+<?xml version="1.0" encoding="UTF-8"?>
```
I did not touch this at all :')

helicopter88 commented:
I'd rather leave the identity handling to github. you can tag people/add more signed-offs if they actually worked with you.
Also, your latest commit is wrong, @AbraaoMota has fixed it however.

Proca-Consul commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,98 @@
+            jsonMicroServiceOutput = PostProcessorUtils.obtainJSONFile(path);
+
+            JSONArray errors;
+            if (!(errors = (JSONArray) jsonMicroServiceOutput.get("errors")).isEmpty()) {
```
The implicit assumption is that micro service generate valid JSON. It must not fail indeed.
Also it is clean enough, what is wrong with it?

mihaicar replied:
do we really need this many empty lines?

Proca-Consul commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,98 @@
+            if (!(errors = (JSONArray) jsonMicroServiceOutput.get("errors")).isEmpty()) {
+                System.out.println("A microservice failed to produce a valid output. " +
+                        "Details on the errors: " + errors.toString());
+                // TODO: decide what to do in this case. For the moment, skipping output.
```
Yes, checked

mihaicar commented on backend/postprocessor/src/main/java/postprocessor/Aggregator.java:
```diff
@@ -0,0 +1,95 @@
+    private String generateJSONOutput(Map<ErrorType, List<JSONObject>> map) {
+        JSONObject finalOutput = new JSONObject();
+        for (ErrorType type : ErrorType.values()) {
+
```
return tag is empty above and again some useless empty lines

helicopter88 commented:
I thought in the commits you were implying that you did something to the class field of the classes, not removing the .class files, my fault.

helicopter88 commented:
LGTM, another approval or two and this should be merged

mikeycar commented on backend/mark_marker/src/main/scala/mark_marker/App.scala:
```diff
@@ -0,0 +1,63 @@
+    case "C" => 55
+    case "D" => 45
+    case "E" => 35
+    case "F" => 25
```
F should be 15

helicopter88 replied:
Will be done

mikeycar commented on backend/mark_marker/src/main/scala/mark_marker/SainsburyException.scala:
```diff
@@ -0,0 +1,3 @@
+package mark_marker
+
+class SainsburyException extends Exception("Unexpected item in bagging area")
```
...

giuliojiang replied:
...

procaconsul replied:
<3

AbraaoMota replied:
As nice as it is, plz remove :)

mikeycar commented on backend/mark_marker/src/main/scala/mark_marker/Utils.scala:
```diff
@@ -0,0 +1,28 @@
+  def matchType(t: String) = t match {
+    case s if s.contains("haskell") => ".hs"
+    case s if s.contains("java") => ".java"
+    case _ => throw new IllegalArgumentException("Invalid language")
```
If we go with the "it's a type, not a language" rhetoric, you should report invalid type i guess?

helicopter88 replied:
Will be changed to "Invalid type: $s"

mikeycar commented on backend/mark_marker/src/main/scala/mark_marker/trainer/InitialTraining.scala:
```diff
@@ -0,0 +1,58 @@
+object InitialTraining {
+
+  private def parse(string: String) = {
+    string.split(":")(1).trim.replace("}", "")
```
I suppose this is what you had to do to get the mark, right?

helicopter88 replied:
Yes as somebody made the training set contain ill-formed JSON

mihaicar reviewed (commented):
Is there any testing evidence of this in action? I'd be interested in looking at that

giuliojiang commented on backend/mark_marker/src/main/scala/mark_marker/trainer/DatabaseConnector.scala:
```diff
@@ -0,0 +1,96 @@
+import edu.stanford.nlp.classify.{Classifier, ColumnDataClassifier}
+
+import scala.io.Source
+
```
could reuse the Java database connector here as well? Should be easy to import it from the infrastructure

helicopter88 replied:
That would make the entire backend depend on infrastructure.
A correct approach would be to move it where it belongs, in lib and both of us just take it from there.

giuliojiang replied:
something for when we have a bit more time then

giuliojiang commented on lib/Makefile:
```diff
@@ -1,17 +1,10 @@
-	mvn clean -q
-	- rm -rf training
+	- rm -rf repo
 
```
cleaning should not remove repo

giuliojiang commented on Makefile:
```diff
@@ -5,23 +5,26 @@ test: all
 	cd backend && mvn -T 1C test -q
 
-infrastructure: simple-haskell-marker
+infrastructure: repo
```
infrastructure DOES need simple-haskell-marker, as the microservice updater is linked in directly

giuliojiang replied:
infrastructure build crashes if .m2 is clean. I'll fix this now

helicopter88 replied:
oh oops I removed it.
It doesn't really depend on it, as with this change, we can just move it to backend and just build it before

giuliojiang commented on lib/Makefile:
```diff
@@ -1,17 +1,10 @@
-
-test:
-	mvn test -q
+all: repo
```
infrastructure/infrastructure/pom.xml should also be edited to point to the new stanfordclassifier repo location

giuliojiang reviewed (commented):
Some comments here.
I'm going to push a little fix to the makefiles to sort out a potential problem with the local maven cache

giuliojiang commented on backend/mark_marker/src/main/scala/mark_marker/trainer/DatabaseConnector.scala:
```diff
@@ -0,0 +1,96 @@
+    sb.toString()
+  }
+
+  def getSerialisedClassifier(exercise: String): Classifier[String, String] = {
```
you are storing the serialized binary classifier in the database. I don't think that's a good approach.
The database should have a table
[exercise_name, source_code]
containing simply the source code for all exercises.
Then the updater will simply have to put a new line to the table containing the new source code for the training set,
and the microservice can run a clock to refresh the local serialized classifiers once very some time.
Having the binary blob in the database means that you would have to update the entire blob every time you add a new source file sample.

It should be 1 line per student source, not 1 line containing the entire dataset

helicopter88 replied:
I wasn't sure whether to store the big training set or just line per line.
We will probably need a separate table to hold the serialized blob however.

giuliojiang replied:
don't hold the serialized blob in the db. Generate that upon server startup.
I'll approve the PR for now, but we definitely should change it

helicopter88 replied:
I would have to add a service/server to keep the thing refreshed
I believe that the serialized classifier would make it smooth enough

giuliojiang replied:
you would still need a service/server to keep the db refreshed, and it's no good if we have many instances of the server running on different locations

helicopter88 replied:
Well to keep the db updated according to feedback I do not need to interact with the marking service, I can just pull/update the training set, retrain and upload the serialized class, and by magic, the ms will use it next time it will run

giuliojiang replied:
you still need to retrain. So instead of retraining and pushing to db, just retrain and save locally. That's it.

giuliojiang commented on backend/mark_marker/src/main/scala/mark_marker/trainer/DatabaseConnector.scala:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  def getSerialisedClassifier(exercise: String): Classifier[String, String] = {
+    val stat = connection.prepareStatement(s"SELECT serialized FROM MarkMarkerTraining WHERE exercise_name = ?")
```
I'll have to remember that we need to only allow users to specify one of these types of exercises in the frontend

AbraaoMota commented on backend/mark_marker/src/main/scala/mark_marker/Utils.scala:
```diff
@@ -0,0 +1,28 @@
+import scala.io.Source
+
+/**
+  * Created by eb1314 on 27/11/16.
```
^

AbraaoMota commented on backend/mark_marker/src/main/scala/mark_marker/trainer/InitialTraining.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  private def analyseFolder(file: File, extension: String) = {
+    if (file.isFile) throw new SainsburyException
```
hmmmm

giuliojiang reviewed (approved):
this was pair-coded, it's good to merge

giuliojiang reviewed (commented):
Just noticed that this is not added to
infrastructure/microservices.py
Please do and test before merging
